rpc/handlers.go

```diff
@@ -1379,6 +1395,82 @@ func (h *Handler) SpecVersion() (string, *jsonrpc.Error) {
 	return "v0.5.0", nil
 }
 
+func randomID() uint64 {
+	var n uint64
+	for err := binary.Read(rand.Reader, binary.LittleEndian, &n); err != nil; {
+	}
+	return n
+}
+
+func (h *Handler) SubscribeNewHeads(ctx context.Context) (uint64, *jsonrpc.Error) {
+	w, ok := jsonrpc.ConnFromContext(ctx)
+	if !ok {
+		return 0, jsonrpc.Err(jsonrpc.MethodNotFound, nil)
+	}
+
+	id := randomID()
+	subscriptionCtx, subscriptionCtxCancel := context.WithCancel(ctx)
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.subscriptions[id] = &subscription{
+		cancel: subscriptionCtxCancel,
+		conn:   w,
+	}
+
+	go func() {
+		sink := make(chan *core.Header)
+		sub := h.bcReader.SubscribeNewHeads(sink)
```

Comment:
This event will be triggered both on new blocks and reorg events. So the same block might be emitted multiple times.
Let's do what ETH does and not notify users when chain reorgs? https://docs.alchemy.com/reference/newheads

> NOTE: Chain Reorganizations (ReOrgs)

> When a chain reorganization occurs, this subscription will emit an event containing all new headers (blocks) for the new chain. This means that you may see multiple headers emitted with the same height (block number), and when this happens the later (highest) block number should be taken as the correct one after a reorganization.